### PR TITLE
fix(router): prevent `RouterLinkActive` from causing an infinite CD loop

### LINKED ISSUE #15825: RouterLinkActive causes `Maximum call stack size` error when wrapped in `ng-template`
**I'm submitting a ...**  (check one with "x")
```
[x] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**
In my application, I'm loading users from a server. Those users are then shown in a searchable navigation list so you can select one and view the details. Something like this.

```html
<search-list [items]="users">
      <ng-template let-user>
            <a [routerLink]="[user.id]">{{ user.name }}</a>
      </ng-template>
</search-list>
```

That `ng-template` is then used in the `search-list` to render each individual user and works perfectly.

Off course I also want to visually show what user is selected by using the `routerLinkActive` attribute. 

```html
<a [routerLink]="[user.id]" routerLinkActive="active">{{ user.name }}</a>
```

And here is where it fails. If you navigated to `users/2` for instance and you refresh the page, this means that the route of user `2` should be set to active after the data comes back from the server. And this is very important! If the data is static and thus not delayed, it works. But if the data comes in later, it will result in the following error.

```
ERROR RangeError: Maximum call stack size exceeded
    at isViewDebugError (core.es5.js:8629)
    at callWithDebugContext (core.es5.js:13020)
    at Object.debugCheckAndUpdateView [as checkAndUpdateView] (core.es5.js:12553)
    at ViewRef_.detectChanges (core.es5.js:10122)
    at RouterLinkActive.update (router.es5.js:5172)
    at RouterLinkActive.ngAfterContentInit (router.es5.js:5137)
    at callProviderLifecycles (core.es5.js:11046)
    at callElementProvidersLifecycles (core.es5.js:11027)
    at callLifecycleHooksChildrenFirst (core.es5.js:11011)
    at checkAndUpdateView (core.es5.js:12028)
    at callViewAction (core.es5.js:12340)
    at execEmbeddedViewsAction (core.es5.js:12312)
    at checkAndUpdateView (core.es5.js:12026)
    at callViewAction (core.es5.js:12340)
    at execEmbeddedViewsAction (core.es5.js:12312)
```

This also only seems to happen if you wrap it inside a `ng-template`, otherwise it works fine if you just render it in a list.

**Expected behavior**
The `routerLinkActive` property should work in a `<ng-template>` as well and mark the link as active even if the data is coming in delayed. It worked in Angular 2.x so it seems a regression bug.

**Minimal reproduction of the problem with instructions**
I've set up a plunkr to showcase what happens: http://plnkr.co/edit/alu9GXQvIaflkTrrIsxl?p=preview

If you run the example, at the top you will have the navigation that is being "loaded". I used a delayed observable to fake that it is coming from a server.

At the bottom you have the content which just renders the `page` parameter of the url to indicate the active page.

In the routing, I set the default path to redirect to the `rainbow` page (as you will notice in the content section because it renders `rainbow`). This step is important because it mimics the user refreshing the application while the active page is `rainbow`.

**What is the motivation / use case for changing the behavior?**
It seems like a regression from Angular 2.x because it worked back then.

**Please tell us about your environment:**
macOS Sierra, VS Code, yarn, WDS

* **Angular version:** 4.0.0 & 4.0.1

* **Browser:** all
 
* **Language:** TypeScript 2.2

* **Node (for AoT issues):** `node --version` =  6.9.1

### PR DESCRIPTION
**Pull request by vicb**
fixes #15825

### CODE REVIEW

File: packages/router/test/regression_integration.spec.ts [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+         const instance = fixture.componentInstance;
+         instance.show = true;
+         expect(() => advance(fixture)).not.toThrow();
+       }));
+  });
+
+});
+
+function advance<T>(fixture: ComponentFixture<T>): void {
```
Comment (jasonaden) [inline packages/router/test/regression_integration.spec.ts]: We already have both these functions in the integration.spec.ts. They should be centralized.

**Review by tbosch — APPROVED** [review]

**Review by jasonaden — COMMENTED** [review]
Approved

**Comment by SamVerschueren** [comment]
Great work @vicb !

**Comment by youngmob** [comment]
is it possible to know when this fix will be released?

**Comment by SamVerschueren** [comment]
It is already in 4.0.3

**Pull request merged**